Mathlib/Topology/Order/LowerTopology.lean

```diff
@@ -0,0 +1,289 @@
+  simp_rw [lowerBasis, coe_upperClosure, compl_unionᵢ]
+  ext s
+  constructor
+  · rintro ⟨F, hF, rfl⟩
+    refine' ⟨(fun a => Ici aᶜ) '' F, ⟨hF.image _, image_subset_iff.2 fun _ _ => ⟨_, rfl⟩⟩, _⟩
+    simp only [interₛ_image]
+  · rintro ⟨F, ⟨hF, hs⟩, rfl⟩
+    haveI := hF.to_subtype
+    rw [subset_def, Subtype.forall'] at hs
+    choose f hf using hs
+    exact ⟨_, finite_range f, by simp_rw [binterᵢ_range, hf, interₛ_eq_interᵢ]⟩
+#align lower_topology.is_topological_basis LowerTopology.isTopologicalBasis
+
+/-- A function `f : β → α` with lower topology in the codomain is continuous provided that the
+preimage of every interval `Set.Ici a` is a closed set.
+
+TODO: upgrade to an `iff`. -/
+lemma continuous_of_Ici [TopologicalSpace β] {f : β → α} (h : ∀ a, IsClosed (f ⁻¹' (Ici a))) :
+    Continuous f := by
+  obtain rfl := LowerTopology.topology_eq α
+  refine continuous_generateFrom ?_
+  rintro _ ⟨a, rfl⟩
+  exact (h a).isOpen_compl
+
+end Preorder
+
+section PartialOrder
+
+variable [PartialOrder α] [TopologicalSpace α] [LowerTopology α]
+
+-- see Note [lower instance priority]
+/-- The lower topology on a partial order is T₀. -/
+instance (priority := 90) toSpace : T0Space α :=
+  (t0Space_iff_inseparable α).2 fun x y h =>
+    Ici_injective <| by simpa only [inseparable_iff_closure_eq, closure_singleton] using h
+
+end PartialOrder
+
+end LowerTopology
+
+instance [Preorder α] [TopologicalSpace α] [LowerTopology α] [OrderBot α] [Preorder β]
+    [TopologicalSpace β] [LowerTopology β] [OrderBot β] : LowerTopology (α × β) where
+  topology_eq_lowerTopology := by
+    refine' le_antisymm (le_generateFrom _) _
+    · rintro _ ⟨x, rfl⟩
+      exact ((LowerTopology.isClosed_Ici _).prod <| LowerTopology.isClosed_Ici _).isOpen_compl
+    rw [(LowerTopology.isTopologicalBasis.prod LowerTopology.isTopologicalBasis).eq_generateFrom,
+      le_generateFrom_iff_subset_isOpen, image2_subset_iff]
+    rintro _ ⟨s, hs, rfl⟩ _ ⟨t, ht, rfl⟩
+    dsimp
+    simp_rw [coe_upperClosure, compl_unionᵢ, prod_eq, preimage_interᵢ, preimage_compl]
+    -- without `let`, `refine` tries to use the product topology and fails
+    let _ : TopologicalSpace (α × β) := generateFrom { s | ∃ a, Ici aᶜ = s }
+    refine (isOpen_binterᵢ hs fun a _ => ?_).inter (isOpen_binterᵢ ht fun b _ => ?_)
+    · exact GenerateOpen.basic _ ⟨(a, ⊥), by simp [Ici_prod_eq, prod_univ]⟩
+    · exact GenerateOpen.basic _ ⟨(⊥, b), by simp [Ici_prod_eq, univ_prod]⟩
+
+section CompleteLattice
+
+variable [CompleteLattice α] [CompleteLattice β] [TopologicalSpace α] [LowerTopology α]
+  [TopologicalSpace β] [LowerTopology β]
+
+theorem InfₛHom.continuous (f : InfₛHom α β) : Continuous f := by
+  refine LowerTopology.continuous_of_Ici fun b => ?_
+  convert LowerTopology.isClosed_Ici (infₛ <| f ⁻¹' Ici b)
+  refine' Subset.antisymm (fun a => infₛ_le) fun a ha => le_trans _ <| OrderHomClass.mono f ha
+  simp [map_infₛ]
+#align Inf_hom.continuous InfₛHom.continuous
```

Comment:
[The explanation of the typeclass issues on zulip.](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Instances.20in.20!4.232138/near/326644735)

Here's one way to hint to Lean which instance we want. Note that we're actually casting `f` to an `OrderHom` now, and `simp` is having the same trouble synthesising the `OrderHomClass` instance so we can't simp `((f : α →ₒ β) : α → β)` down to just `f : α → β`, so I use `refine` instead.
```suggestion
theorem InfₛHom.continuous (f : InfₛHom α β) : Continuous f := by
  refine LowerTopology.continuous_of_Ici fun b => ?_
  convert LowerTopology.isClosed_Ici (infₛ <| f ⁻¹' Ici b)
  refine' Subset.antisymm (fun a => infₛ_le) fun a ha => le_trans _ <|
    OrderHomClass.mono (f : α →o β) ha
  refine' LE.le.trans _ (map_infₛ f _).ge
  simp
#align Inf_hom.continuous InfₛHom.continuous
```